entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -89,14 +97,54 @@ public void execute(EntityQueryRequest request, StreamObserver<ResultSetChunk> r
         .convertToEDSQuery(request, attrNameToEDSAttrMap.get(request.getEntityType()));
     Iterator<Document> documentIterator = entitiesCollection.search(
         DocStoreConverter.transform(tenantId.get(), query));
-    List<Entity> entities = convertDocsToEntities(documentIterator);
-
-    //Build result
-    //TODO : chunk response. For now sending everything in one chunk
-    responseObserver.onNext(convertEntitiesToResultSetChunk(
-        entities,
-        request.getSelectionList(),
-        attrNameToEDSAttrMap.get(request.getEntityType())));
+
+    if (!documentIterator.hasNext()) {
+      ResultSetChunk.Builder resultBuilder = ResultSetChunk.newBuilder();
+      resultBuilder.setResultSetMetadata(ResultSetMetadata.newBuilder()
+          .addAllColumnMetadata(
+              () -> request.getSelectionList().stream().map(Expression::getColumnIdentifier)
+                  .map(
+                      ColumnIdentifier::getColumnName)
+                  .map(s -> ColumnMetadata.newBuilder().setColumnName(s).build()).iterator())
+          .build());
+      resultBuilder.setIsLastChunk(true);
+      resultBuilder.setChunkId(1);
+      responseObserver.onNext(resultBuilder.build());
+    }
+    boolean isNewChunk = true;
+    int chunkId = 1, rowCount = 0;
+    ResultSetChunk.Builder resultBuilder = ResultSetChunk.newBuilder();
+    while (documentIterator.hasNext()) {
+      Optional<Entity> entity = DOCUMENT_PARSER.parseOrLog(documentIterator.next(), Entity.newBuilder());
+      // Set metadata for new chunk
+      if (isNewChunk) {
+        resultBuilder.setResultSetMetadata(ResultSetMetadata.newBuilder()
+            .addAllColumnMetadata(
+                () -> request.getSelectionList().stream().map(Expression::getColumnIdentifier)
+                    .map(
+                        ColumnIdentifier::getColumnName)
+                    .map(s -> ColumnMetadata.newBuilder().setColumnName(s).build()).iterator())
+            .build());
+        isNewChunk = false;
+      }
+      if (entity.isPresent()) {
```

Comment:
style nit - `entity.ifPresent(entity => ...`

Reply:
Skipping this, otherwise I'll need to make the variables used in lambda final.

Reply:
yes, that's another one of the benefits 😉